tutorial/cytoscape/utils.py

```diff
@@ -0,0 +1,104 @@
+from textwrap import dedent
+
+import dash_core_components as dcc
+import dash_html_components as html
+from tutorial import styles
+
+
+
+def CreateDisplay(scope):
+    def Display(example_string):
+        return html.Div([
+            dcc.SyntaxHighlighter(
+                dedent(example_string).strip(),
+                language='python',
+                customStyle={'marginBottom': 10, 'borderLeft': 'thin #C8D4E3 solid'}
+            ),
+            html.Div(
+                children=eval(dedent(example_string), scope),
+                style={
+                    'border': 'thin lightgrey solid',
+                    'margin-top': '-10px',
+                    'padding': '15px'
+                }
+            )
+        ])
+    return Display
+
+
+def merge(*args):
+    merged = {}
+    for arg in args:
+        for key in arg:
+            merged[key] = arg[key]
+    return merged
+
+
+def section_title(title):
+    return html.H4(title, style={"marginTop": "20px"})
+
+
+def html_table(
+    df,
+    base_column_style={},
+    table_style={},
+    column_style={},
+    cell_style={},
+    cell_style_by_column={},
+    header_style={},
+    header_style_by_column={},
+    row_style={},
+    row_style_by_index={},
+    odd_row_style={},
+    conditional_cell_style=(lambda cell, column: {})
+):
+    header = []
+    for column in df.columns:
+        header.append(
+            html.Th(
+                column,
+                style=merge(
+                    row_style,
+                    base_column_style,
+                    cell_style,
+                    column_style.get(column, {}),
+                    header_style,
+                    header_style_by_column.get(column, {}),
+                ),
+            )
+        )
+
+    rows = []
+    for i in range(len(df)):
+        row = []
+        for column in df.columns:
+            row.append(
+                html.Td(
+                    df.iloc[i][column],
```

Comment:
```suggestion
                    df.loc[i, column],
```

Otherwise you are doing chained indexing, which is not a good practice.
Reference: https://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
(I once covered this in an lesson introducing Pandas: https://github.com/mkcor/advanced-pandas)

Reply:
Thanks for the heads up! This is an excellent idea, however I realized I'm not using this function anywhere in the Cytoscape docs, so I think I could just remove this function. I'll keep this in mind tho!